src/login/mod.rs

```diff
@@ -1,104 +1,220 @@
     if !browser_permission {
         anyhow::bail!("In order to log in you must allow Wrangler to open your browser. If you don't want to do this consider using `wrangler config`");
     }
 
-    open_browser(&format!(
-        "https://dash.cloudflare.com/wrangler?key={0}",
-        pubkey_encoded
-    ))?;
-
-    let encrypted_token_str = poll_token(pubkey_filtered)?;
-    let encrypted_token = base64::decode_block(encrypted_token_str.as_str())?;
-    let mut token_bytes: [u8; 128] = [0; 128];
-
-    rsa.private_decrypt(&encrypted_token, &mut token_bytes, Padding::PKCS1)?;
-    let token = str::from_utf8(&token_bytes)?.trim_matches(char::from(0));
+    if wsl::is_wsl() {
+        open_browser(auth_url.as_str(), true)?;
+    } else {
+        open_browser(auth_url.as_str(), false)?;
+    }
 
-    let user = GlobalUser::TokenAuth {
-        api_token: token.to_string(),
+    // Get authorization code and CSRF state from local HTTP server
+    let params_response = match block_on(http_server_get_params()) {
+        Ok(params) => params,
+        Err(_) => anyhow::bail!("Failed to receive authorization code from local HTTP server"),
     };
-    global_config(&user, true)?;
-
-    Ok(())
-}
+    let params_values: Vec<&str> = params_response.split_whitespace().collect();
+    if params_values.is_empty() {
+        anyhow::bail!("Failed to receive authorization code from local HTTP server")
+    }
 
-#[derive(Deserialize)]
-struct TokenResponse {
-    result: String,
-}
+    // Check if user has given consent, or if an error has been encountered
+    let response_status = params_values[0];
+    if response_status == "denied" {
+        anyhow::bail!("Consent denied. You must grant consent to Wrangler in order to login. If you don't want to do this consider using `wrangler config`")
+    } else if response_status == "error" {
+        anyhow::bail!("Failed to receive authorization code from local HTTP server")
+    }
 
-/// Poll for token, bail after 500 seconds.
-fn poll_token(token_id: String) -> Result<String> {
-    let mut request_params = HashMap::new();
-    request_params.insert("token-id", token_id);
+    // Get authorization code and CSRF state
+    if params_values.len() != 3 {
+        anyhow::bail!(
+            "Failed to receive authorization code and/or csrf state from local HTTP server"
+        )
+    }
 
-    let client = reqwest::blocking::Client::builder().build()?;
-    let timer = Timer::new().interval_ms(1000).iter();
+    let auth_code = params_values[1];
+    let recv_csrf_state = params_values[2];
+
+    // Check CSRF token to ensure redirect is legit
+    let recv_csrf_state = CsrfToken::new(recv_csrf_state.to_string());
+    if recv_csrf_state.secret() != csrf_state.secret() {
+        anyhow::bail!(
+            "Redirect URI CSRF state check failed. Received: {}, expected: {}",
+            recv_csrf_state.secret(),
+            csrf_state.secret()
```

Comment:
Are these values safe to print? I honestly have no idea, so just checking :)

Reply:
Nice catch! I think it's possible for someone to use the previous `printed URL + the expected csrf_state` to get an authorization grant. Better be safe and remove it